Client/macos/MacOSBluetoothConnector.mm

```diff
@@ -0,0 +1,162 @@
+-(void)rfcommChannelOpenComplete:(IOBluetoothRFCOMMChannel *)rfcommChannel status:(IOReturn)error
+{
+    
+    if ( error != kIOReturnSuccess ) {
+        fprintf(stderr,"Error - could not open the RFCOMM channel. Error code = %08x.\n",error);
+        return;
+    }
+    else{
+        fprintf(stderr,"Connected. Yeah!\n");
+    }
+    
+}
+// this function fires when the channel receives data
+-(void)rfcommChannelData:(IOBluetoothRFCOMMChannel *)rfcommChannel data:(void *)dataPointer length:(size_t)dataLength
+{
+    NSString  *message = [[NSString alloc] initWithBytes:dataPointer length:dataLength encoding:NSUTF8StringEncoding];
+    delegateCPP->dataRec([message UTF8String]);
+}
+
+
+@end
+
+int MacOSBluetoothConnector::send(char* buf, size_t length)
+{
+    fprintf(stderr,"Sending Message\n");
+	//write buffer to channel
+    [(__bridge IOBluetoothRFCOMMChannel*)rfcommchannel writeSync:(void*)buf length:length];
+	return 1; // what to return here?
+}
+
+
+void MacOSBluetoothConnector::connectToMac(MacOSBluetoothConnector* MacOSBluetoothConnector)
+{
+	MacOSBluetoothConnector->running = 1;
+	//get device
+	IOBluetoothDevice *device = (__bridge IOBluetoothDevice *)MacOSBluetoothConnector->rfcommDevice;
+	// create new channel
+	IOBluetoothRFCOMMChannel *channel = [[IOBluetoothRFCOMMChannel alloc] init];
+	// create sppServiceid
+	IOBluetoothSDPUUID *sppServiceUUID = [IOBluetoothSDPUUID uuid16: kBluetoothSDPUUID16RFCOMM];
+	// get sppServiceRecord
+    IOBluetoothSDPServiceRecord *sppServiceRecord = [device getServiceRecordForUUID:sppServiceUUID];
+    // get rfcommChannelID from sppServiceRecord
+	UInt8 rfcommChannelID;
+    [sppServiceRecord getRFCOMMChannelID:&rfcommChannelID];
+	// setup delegate
+	AsyncCommDelegate* asyncCommDelegate = [[AsyncCommDelegate alloc] init];
+	asyncCommDelegate->delegateCPP = MacOSBluetoothConnector;
+	// try to open channel
+	if ( [device openRFCOMMChannelAsync:&channel withChannelID:rfcommChannelID delegate:asyncCommDelegate] != kIOReturnSuccess ) {
+		throw "Error - could not open the rfcomm.\n";
+	}
+	// store the channel
+	MacOSBluetoothConnector->rfcommchannel = (__bridge void*) channel;
+	
+	printf("Successfully connected");
+	// keep thread running
+	while (MacOSBluetoothConnector->running) {
+		[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
+	}
+}
+
+void MacOSBluetoothConnector::connect(const std::string& addrStr){
+	// convert mac adress to nsstring
+	NSString *addressNSString = [NSString stringWithCString:addrStr.c_str() encoding:[NSString defaultCStringEncoding]];
+	// get device based on mac adress
+	IOBluetoothDevice *device = [IOBluetoothDevice deviceWithAddressString:addressNSString];
+	// if device is not connected
+	if (![device isConnected]) {
+		[device openConnection];
+	}
+	// store the device in an variable
+	rfcommDevice = (__bridge void*) device;
+	uthread = new std::thread(MacOSBluetoothConnector::connectToMac, this);
```

Comment:
no need to run a new thread, those ops already run in a separate thread. you can block here

Please see [IBluetoothConnector.h](https://github.com/Plutoberth/SonyHeadphonesClient/blob/master/Client/IBluetoothConnector.h)

Reply:
hmmm, if I try to keep the thread running (to avoid the channel to be closed), then the disconnect function call doesn't work anymore. Maybe a new thread is still needed?

Reply:
I'm not sure what then rest of the code in `connectToMac` does, but at the end you block the thread forever, so in the example that you sent me the operation will never be finished.

```
	// keep thread running
	while (MacOSBluetoothConnector->running) {
		[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
	}
```

This makes the thread that I create in `CrossPlatformGUI.cpp` never return (setFromAsync runs the lambda in a new thread), so the "disconnect" button is not shown.

```
this->_connectFuture.setFromAsync([this]() { this->_bt.connect(this->_connectedDevice.mac); });
```

Could you explain why `AsyncCommDelegate` is needed? I think that this function could be erased, and everything would be fine. I made it so the functions that need to block (send,recv,connect) can block freely, so there's no need for async stuff.

Reply:
I'm not sure what then rest of the code in `connectToMac` does, but at the end you block the thread forever, so in the example that you sent me the operation will never be finished.

```
	// keep thread running
	while (MacOSBluetoothConnector->running) {
		[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1]];
	}
```

This makes the thread that I create in CrossPlatformGUI.cpp never return so the program waits and the "disconnect" button is not shown.

Could you explain why `AsyncCommDelegate` is needed? I think that this function could be erased, and everything would be fine. I made it so the functions that need to block (send,recv,connect) can block freely, so there's no need for async stuff.

Reply:
The asyncCommDelegate is used for the the callback functions, such as `rfcommChannelOpenComplete` and `rfcommChannelData` for when it receives data. I'm not very experienced in obj-c++ so I don't know if it can be done differently.

Reply:
If `openRFCOMMChannelAsync` is available, then surely `openRFCOMMChannelSync` is too - https://developer.apple.com/documentation/iobluetooth/iobluetoothdevice/1431606-openrfcommchannelsync . The sync API is probably simpler and will allow us to reduce the amount of errors.

Reply:
I tried it but it doesn't seem to solve the issues :( 
Now it just crashes, and gives `NULLing out io_service_t.` error.